core/dbt/adapters/materialization/factory.py

```diff
@@ -0,0 +1,39 @@
+from typing import Dict, Optional, Type
+
+from dbt.adapters.materialization import models
+from dbt.adapters.relation import RelationFactory
+from dbt.adapters.relation.models import RelationRef
+
+
+class MaterializationFactory:
+    def __init__(
+        self,
+        relation_factory: RelationFactory,
+        materialization_map: Optional[
+            Dict[models.MaterializationType, Type[models.Materialization]]
+        ] = None,
+    ):
+        self.relation_factory = relation_factory
+        self.materialization_map = materialization_map or {
+            models.MaterializationType.MaterializedView: models.MaterializedViewMaterialization
+        }
+
+    def make_from_runtime_config(
+        self,
+        runtime_config,
+        materialization_type: models.MaterializationType,
+        existing_relation_ref: Optional[RelationRef] = None,
+    ) -> Optional[models.Materialization]:
+        if parser := self._get_parser(materialization_type):
+            assert self.relation_factory is not None  # mypy
+            return parser.from_runtime_config(
+                runtime_config=runtime_config,
+                relation_factory=self.relation_factory,
+                existing_relation_ref=existing_relation_ref,
+            )
+        return None
+
+    def _get_parser(
```

Comment:
why is this called _get_parser when it is returning an `Optional[Type[models.Materialization]]`

Reply:
I didn't want to confuse it with getting an instance of `Materialization`. It's returning the appropriate class for the type. And the class, at the end of the day, is really just a parser. I'm not tied to it, but I didn't want to call it `get_materialization` since it doesn't actually produce a `Materialization` instance.

Reply:
Renaming this `_get_materialization_class` and the counterpart in `RelationFactory` `_get_relation_class`.